### PR TITLE
Fix headless

### PR DESCRIPTION
**Pull request by nullkiller**
It does not looks like a hotfix so will be retargeted later to develop
Fixes headless mode and a set of fixes for AI. So far:
- attempt to fix parallel write access to hero roles map
- fix counting mana cost of Fly a few times
- added value for pyramid and cyclops stockpile

### CODE REVIEW

**Comment by DjWarmonger** [comment]
AI should already consider Stockpile for its resources, and I saw it attack Stockpile at least one. Still very rare and possible only when there are no better targets.

Also there is no point to hardcode Pyramid since we already have common interface for Banks.

**Comment by nullkiller** [comment]
> AI should already consider Stockpile for its resources, and I saw it attack Stockpile at least one. Still very rare and possible only when there are no better targets.
> 
> Also there is no point to hardcode Pyramid since we already have common interface for Banks.

Yes it will if there is a need for such resources. In general there is always such need but may be of different strength so there are cases when there is no need and priority will be 0. Now it will have small priority and in case there is no other targets it will still attack stockpile.

**Comment by nullkiller** [comment]
Pyramid is hardcoded because it contains spells. My generic interface towards banks can not handle spells

File: AI/Nullkiller/Pathfinding/AINodeStorage.h [hunk 1]
```diff
@@ -41,11 +41,19 @@ namespace AIPathfinding
 	const int CHAIN_MAX_DEPTH = 4;
 }
 
+enum DayFlags : ui8
+{
+	NONE = 0,
+	FLY_CASTED = 1,
+	WATER_WALK_CASTED = 2
```
Comment (Alexander-Wilms) [inline AI/Nullkiller/Pathfinding/AINodeStorage.h:48]:
```suggestion
	FLY_CAST = 1,
	WATER_WALK_CAST = 2
```

**Pull request merged**